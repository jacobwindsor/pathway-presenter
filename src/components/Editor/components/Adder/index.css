.adder-wrapper {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    padding: 1.6rem;
}

.adder-wrapper .header {
    text-align: center;
    color: white;
    background-color: #008ba3 !important; /* Cyan500 dark */
    margin: -1.6rem;
    margin-bottom: 1.6rem;
    padding: 1.6rem;
    padding-bottom: calc(1.6rem + 32px);
    position: relative;
}

.adder-wrapper .header a {
    color: white !important;
}

.adder-wrapper .header .logo-wrapper {
    background-color: #ff4081;
    border-radius: 50%;
    width: 64px;
    height:64px;
    padding: 1rem;
    position: absolute;
    left: 50%;
    transform: translate(-50%, 8px);
}

.adder-wrapper .content {
    margin-top: calc(1.6rem + 32px);
}

.adder-wrapper .content .create-button {
    margin-top: .8rem !important;
}
