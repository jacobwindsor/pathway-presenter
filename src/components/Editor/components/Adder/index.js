import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Paper from 'material-ui/Paper';
import TextField from 'material-ui/TextField';
import FlatButton from 'material-ui/FlatButton';
import AutoComplete from 'material-ui/AutoComplete';
import presentations from '../../../../data/presentations';
import { cloneDeep } from 'lodash';
import LogoWhite from '../../../../assets/logo-white.svg';
import './index.css';
import Formsy from 'formsy-react';
import { FormsyText, FormsyAutoComplete } from 'formsy-material-ui/lib';

class Adder extends Component {
  constructor(props) {
    super(props);

    this.state = {
      wpId: '',
      dataSource: [],
      title: '',
      authorName: '',
      version: ''
    };
  }

  componentDidMount() {
    this._isMounted = true;

    presentations
      .list()
      .then(res => {
        if (!this._isMounted) throw new Error('Not mounted');
        return res;
      })
      .then(res =>
        this.setState({
          dataSource: res.map(single => {
            return {
              text: single.title,
              value: single.id
            };
          })
        })
      )
      .catch(err => console.log(err));
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  handleSubmit = data => {
    const { handleCreate } = this.props;
    const copy = cloneDeep(data);
    copy.version = copy.version ? parseInt(copy.version, 10) : 0;
    handleCreate(copy);
  };

  handleSelect = (chosenRequest, index) => {
    const { handleSelect } = this.props;
    const { dataSource } = this.state;
    if (index < 0) return;
    handleSelect(dataSource[index].value);
  };

  render() {
    const errorMessages = {
      select: 'Select a valid presentation',
      title: 'You must enter a valid title',
      authorName: 'You must enter your name(s)',
      wpId: 'You must provide the WikiPathways ID',
      version: 'You can only enter numeric values'
    };

    return (
      <Paper zDepth={1} className="adder-wrapper">
        <div className="header">
          <h1>Pathway Presenter</h1>
          <p>
            Create interactive presentations from pathways on{' '}
            <a
              href="http://wikipathways.org"
              target="_blank"
              rel="noopener noreferrer"
            >
              WikiPathways
            </a>.
          </p>
          <div className="logo-wrapper">
            <img src={LogoWhite} alt="Logo" />
          </div>
        </div>
        <div className="content">
          <h3>Select a Pathway Presentation</h3>
          <AutoComplete
            filter={AutoComplete.fuzzyFilter}
            maxSearchResults={5}
            fullWidth={true}
            openOnFocus={true}
            hintText={'Search by title'}
            dataSource={this.state.dataSource}
            onNewRequest={this.handleSelect}
          />
          <Formsy.Form
            onValid={() => this.setState({ canSubmit: true })}
            onInvalid={() => this.setState({ canSubmit: false })}
            onValidSubmit={this.handleSubmit}
            onInvalidSubmit={() => null}
          >
            <h3 id="creator-form-title">Create a Pathway Presentation</h3>
            <FormsyText
              name="title"
              required
              requiredError={errorMessages.title}
              hintText="The TCA Cycle"
              floatingLabelText="Title"
              fullWidth
              updateImmediately
            />
            <FormsyText
              required
              name="authorName"
              requiredError={'You must fill this in'}
              validationError={errorMessages.authorName}
              hintText="Gregor Mendel, Frederick Sanger"
              floatingLabelText="Author Name(s)"
              fullWidth
              updateImmediately
            />
            <FormsyText
              name="wpId"
              required
              validationError={errorMessages.wpId}
              validationErrors={{
                beginsWithWP:
                  'The WikiPathways ID must begin with WP followed by a number'
              }}
              validations={{
                beginsWithWP: (values, value) => {
                  const beginning = value.slice(0, 2);
                  const end = value.slice(2);
                  return beginning === 'WP' && !isNaN(end);
                }
              }}
              hintText="WP78"
              floatingLabelText="WikiPathways ID"
              value={''}
              fullWidth
              updateImmediately
            />
            <FormsyText
              name="version"
              validations="isNumeric"
              validationError={errorMessages.version}
              hintText="Type '0' for latest"
              floatingLabelText="WikiPathways Version (optional)"
              fullWidth
              updateImmediately
            />
            <FlatButton
              label="Create"
              primary={true}
              type="submit"
              fullWidth={true}
              className="create-button"
              disabled={!this.state.canSubmit}
            />
          </Formsy.Form>
        </div>
      </Paper>
    );
  }
}

Adder.propTypes = {
  handleCreate: PropTypes.func.isRequired,
  handleSelect: PropTypes.func.isRequired
};

export default Adder;
