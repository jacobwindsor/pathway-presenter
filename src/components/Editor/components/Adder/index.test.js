import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adder from './index';

describe('The Adder. Shows a form to create a new presentation', () => {
  it('should render correctly', () => {
    const comp = shallow(
      <Adder handleCreate={() => null} handleSelect={() => null} />
    );
    expect(toJson(comp)).toMatchSnapshot();
  });

  it('should fire handleCreate', () => {
    const firer = jest.fn();
    const comp = shallow(
      <Adder handleCreate={firer} handleSelect={() => null} />
    );
    comp.find('.create-button').simulate('touchTap');
    expect(firer).toHaveBeenCalledTimes(1);
  });
});
