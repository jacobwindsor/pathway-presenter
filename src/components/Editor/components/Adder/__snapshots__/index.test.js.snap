// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Adder. Shows a form to create a new presentation should render correctly 1`] = `
<Paper
  circle={false}
  className="adder-wrapper"
  rounded={true}
  transitionEnabled={true}
  zDepth={1}
>
  <div
    className="header"
  >
    <h1>
      Pathway Presenter
    </h1>
    <p>
      Create interactive presentations from pathways on
       
      <a
        href="http://wikipathways.org"
        rel="noopener noreferrer"
        target="_blank"
      >
        WikiPathways
      </a>
      .
    </p>
    <div
      className="logo-wrapper"
    >
      <img
        alt="Logo"
        src="logo-white.svg"
      />
    </div>
  </div>
  <div
    className="content"
  >
    <h3>
      Select a Pathway Presentation
    </h3>
    <AutoComplete
      anchorOrigin={
        Object {
          "horizontal": "left",
          "vertical": "bottom",
        }
      }
      animated={true}
      dataSource={Array []}
      dataSourceConfig={
        Object {
          "text": "text",
          "value": "value",
        }
      }
      disableFocusRipple={true}
      filter={[Function]}
      fullWidth={true}
      hintText="Search by title"
      maxSearchResults={5}
      menuCloseDelay={300}
      onNewRequest={[Function]}
      onUpdateInput={[Function]}
      open={false}
      openOnFocus={true}
      targetOrigin={
        Object {
          "horizontal": "left",
          "vertical": "top",
        }
      }
    />
    <h3>
      Create a Pathway Presentation
    </h3>
    <TextField
      disabled={false}
      floatingLabelFixed={false}
      floatingLabelText="Title"
      fullWidth={true}
      hintText="The TCA Cycle"
      multiLine={false}
      onChange={[Function]}
      rows={1}
      type="text"
      underlineShow={true}
      value=""
    />
    <TextField
      disabled={false}
      floatingLabelFixed={false}
      floatingLabelText="Author Name(s)"
      fullWidth={true}
      hintText="Gregor Mendel, Frederick Sanger"
      multiLine={false}
      onChange={[Function]}
      rows={1}
      type="text"
      underlineShow={true}
      value=""
    />
    <TextField
      disabled={false}
      floatingLabelFixed={false}
      floatingLabelText="WikiPathways ID"
      fullWidth={true}
      hintText="WP78"
      multiLine={false}
      onChange={[Function]}
      rows={1}
      type="text"
      underlineShow={true}
      value=""
    />
    <TextField
      disabled={false}
      floatingLabelFixed={false}
      floatingLabelText="WikiPathways Version"
      fullWidth={true}
      hintText="Type '0' for latest"
      multiLine={false}
      onChange={[Function]}
      rows={1}
      type="text"
      underlineShow={true}
      value=""
    />
    <FlatButton
      className="create-button"
      disabled={false}
      fullWidth={true}
      label="Create"
      labelPosition="after"
      labelStyle={Object {}}
      onKeyboardFocus={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={[Function]}
      onTouchTap={[Function]}
      primary={true}
      secondary={false}
    />
  </div>
</Paper>
`;
